services/QuillLMS/app/services/diagnostic_insights/history_by_classroom.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+module DiagnosticInsights
+  class HistoryByClassroom < ApplicationService
+    attr_reader :classroom_id, :ell, :start_date, :end_date
+
+    # Don't look earlier than the start of the 2023-2024 school year
+    DEFAULT_START_DATE = School.school_year_start(DateTime.new(2023, 8, 1))
+
+    STARTER_DIAGNOSTIC = 'starter'
+    INTERMEDIATE_DIAGNOSTIC = 'intermediate'
+    ADVANCED_DIAGNOSTIC = 'advanced'
+
+    def initialize(classroom_id, ell: false, start_date: nil, end_date: nil)
+      @classroom_id = classroom_id
+      @ell = ell
+      @start_date = start_date || DEFAULT_START_DATE
+      @end_date = end_date || Float::INFINITY
+    end
+
+    def run
```

Comment:
You could get rid of the passthrough method `transformed_data` and introduce some [then](https://til.hashrocket.com/posts/f4agttd8si-chaining-then-in-ruby-26) calls for method chaining:

```
def run
  serialized_data
    .then(&method(:group_by_user_id))
    .then(&method(:group_by_diagnostic_type))
    .then(&method(:attach_name_key))
end
```